SDK: improve image building doc, add SDK bootstrapping (#160)
kinvolk/flatcar-docs: docs/reference/developer-guides/sdk-modifying-flatcar.md

Direct questions and suggestions to the [IRC channel][irc] or [mailing list][flatcar-dev].

## Getting started

You’ll need the latest Flatcar SDK installed before we get to the fun part. This is discussed below. After you're set up with an SDK we’ll also do a full image build so you start from a known-good, working Flatcar image. So let's get set up with an SDK chroot and build a bootable image of Flatcar Container Linux. The SDK chroot has a full toolchain and isolates the build process from quirks and differences between host OSes. The SDK must be run on an x86-64 Linux machine, the distro should not matter (Ubuntu, Fedora, etc).

### Prerequisites

**System requirements to get started**

* curl
* git
* bzip2
* gpg
* sudo

Optionally:
* golang if you want to compile cork yourself (see below)

**You also need a proper git setup**

```shell
$ git config --global user.email "you@example.com"
$ git config --global user.name "Your Name"
```

**NOTE**: Do the git configuration as a normal user and not with sudo.

**Finally, you'll need an ssh-agent set up**

> Make sure to set up ssh-agent, otherwise github will return access denied even for our public repos. In most cases this is not an issue as an ssh-agent is enabled by default in most distros. Tje ssh-agent is required as the chroot will access your SSH keys through the agent’s socket to authenticate github access.

Reviewer: typo _Tje_

         | jq -r 'to_entries[] | "\(.value | .channel) tag: v\(.key) - kernel \(.value | .major_software.kernel[0]), released \(.value | .release_date)"' \
         | grep 'alpha tag'
```

Then you can use the "alpha tag" value (without leading 'v') with cork's `--manifest-branch flatcar-alpha-[VERSION]` option to install that SDK version. E.g. to install Alpha release 2801.0.0, run

```shell
$ cork create --verbose --manifest-branch flatcar-alpha-2801.0.0    # Be prepared: This will request root permisions via sudo after downoad!
$ cork enter   # This will also request root permisions via sudo
```

Note that we're using the `--verbose` flag with `cork create` so you get a progress bar for downloading the SDK chroot tarball. The tarball is ~1.4GB in size, so the download can take some time. After the download finished, cork will install the SDK chroot - at that point it will require `sudo` and might prompt for your user password. Note that this prompt can time out if you weren't looking (since the download can take time). Should that happen, run
```shell
$ rm -rf chroot
$ cork create ...   # same options as above, but it will use the cached SDK (i.e. not download again)
```

> As discussed above you most probably want to base your work on the latest Alpah release. Technically, you could use any release - for instance, if you want to fix a bug in Beta or Stable - by using the appropriate `--manifest-branch` option. For releases other than Aplha, you would additionally want to set the `--manifest-name` parameter to `maintenance.xml`, e.g.

Reviewer: ```suggestion
As discussed above you most probably want to base your work on the latest Alpha release. Technically, you could use any release - for instance, if you want to fix a bug in Beta or Stable - by using the appropriate `--manifest-branch` option. Basing your work on the development branches for filing PRs to `main` or the `flatcar-MAJOR` maintenance branches is not covered here. For the given release manifest the SDK checks out the right git branches with matching ebuild files and you should not change them unless you follow the instructions on setting up the nightly binary package URLs.
```
Choosing `maintenance.xml` means that a development branch is used again which I thought we don't want to recommend anymore.

```shell

> $ cork create --verbose --manifest-branch flatcar-stable-2765.2.0  --manifest-name maintenance.xml

Reviewer: ```suggestion
$ cork create --verbose --manifest-branch flatcar-stable-2765.2.0
```

> As discussed above you most probably want to base your work on the latest Alpha release. Technically, you could use any release - for instance, if you want to fix a bug in Beta or Stable - by using the appropriate `--manifest-branch` option. Basing your work on the development branches for filing PRs to `main` or the `flatcar-MAJOR` maintenance branches is not covered here. For the given release manifest the SDK checks out the right git branches with matching ebuild files and you should not change them unless you follow the instructions on setting up the nightly binary package URLs.

Reviewer: ```suggestion
As discussed above you most probably want to base your work on the latest Alpha release. Technically, you could use any release - for instance, if you want to fix a bug in Beta or Stable - by using the appropriate `--manifest-branch` option. Basing your work on the development branches for filing PRs to `main` or the `flatcar-MAJOR` maintenance branches is not covered here but under [the section for using nightly packages][sdktips]. For the given release manifest the SDK checks out the right git branches with matching ebuild files and you should not change them unless you follow the instructions on setting up the nightly binary package URLs.
```

```
to base your work on the Stable 2765.2.0 release published on March 3rd, 2021.

### Enter the SDK

Run
```shell
$ cork enter
```

Verify you are in the SDK chroot:

```shell
$ grep NAME /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
```

To leave the SDK chroot, simply run `exit`.

To use the SDK chroot in the future, run `cork enter` from the directory you installed the SDK to (i.e. where you ran `cork create`).

## Building an image

### A brief introduction to Gentoo

> Flatcar Container Linux is based on ChromionOS, which is based on Gentoo. While the ChromiumOS heritage has faded and is barely visible nowadays, we heavily leverage Gentoo processes and tools.

Reviewer: ```suggestion
Flatcar Container Linux is based on ChromiumOS, which is based on Gentoo. While the ChromiumOS heritage has faded and is barely visible nowadays, we heavily leverage Gentoo processes and tools.
```

It is one of the three repositories that define a Flatcar build:
1. flatcar-scripts (the directory you're in) contains high-level build scripts to build all packages for an image, to build an image, and to bootstrap an SDK.
2. `~/trunk/src/third_party/portage-stable` contains ebuild files of all packages close to (or identical to) Gentoo upstream.
3. `~/trunk/src/third_party/coreos-overlay` contains Flatcar specific packages like ignition and mayday, as well as Gentoo packages which were significantly modified for Flatcar.

The SDK chroot you just entered is self-sustained and has all necessary "host" binaries installed to build Flatcar packages. Flatcar OS image packages are "cross-compiled" (even when host machine equals target machine, e.g. buidling an x86 image on an x86 host). The OS image packages have their own root within the SDK - `/build/amd64-usr/` for x86_64 and `/build/arm64-usr` for ARM64, depending on which _board_ you initialised (see below).

In other words, you'll be dealing with two levels of chroot:
* The top-level SDK chroot, activated by `cork enter`, consisting of the build chroot with ebuild file sources (`src/` in the SDK directory) mounted in

> * The image chroot, in `/]build/>arch>` in the SDK

Reviewer: ```suggestion
* The image chroot, in `/build/<arch>` in the SDK
```

$ mv cork ~/.local/bin
$ export PATH=$PATH:$HOME/.local/bin
```

You may want to add the `PATH` export to your shell profile (e.g. `.bashrc`).

**Alternatively, clone the mantle repo and build cork from sources**

If you want to build cork from sources instead of using an official release, you'll need golang installed as mantle is written in go.

```shell
$ git clone git@github.com:flatcar-linux/mantle.git
$ cd mantle
$ ./build cork
```

Now you want to make the resulting `bin/cork` binary part of your `$PATH`. Some prefer a symlink `~/.local/bin/cork -> ~/code/mantle/bin/cork ` ( `~/.local/bin/` often is in `$PATH` already as per `.bashrc`) - but you’re free to make your own arrangements as you see fit. Simply running `export PATH="$PATH:$(pwd/bin)"` in the mantle repo directory works (for the current shell), too.

Next, use the cork utility to create a project directory. This will hold all of your git repos and the SDK chroot. For building all binaries and images, about 20 gigabytes of free disk space are recommended.

**NOTE** that you can use multiple SDKs in separate directories (and separate shells, of course) at the same time.

```shell
$ mkdir flatcar-sdk
$ cd flatcar-sdk

> $ cork create --verbose # This will request root permisions via sudo

Reviewer: This will check out the `main` branches and use the latest Alpha SDK and without a `set_version` invocation cause people to rebuild everything that was changed since we branched away…

Author: OK, understood.
I'd like the howto to point users to the latest Alpha release (SDK and OS image packages) by default. This way they can base their own work on something known-good instead of a nightly, which may contain breakage.

Does this mean we should always specify `cork create --manifest-branch flatcar-alpha-xxxx` in the above?

$ cork enter # This will request root permisions via sudo
```

Note that we're using the `--verbose` flag with `cork create` so you get a progress bar for downloading the SDK chroot tarball. The tarball is ~1.4GB in size, so the download can take some time.

By default `cork create` will download the latest SDK used for the Alpha release. You can download arbitrary SDKs by using the `--sdk-version vX.Y.Z` command line option to `cork create`. Note that in general, new SDKs are only released for Alpha versions (Beta and Stable releases following Alphas will use the same SDK).Get a list of all Alpha releases by issuing

```shell
$ curl -s https://kinvolk.io/flatcar-container-linux/releases-json/releases.json \
         | jq -r 'to_entries[] | "\(.value | .channel) tag: v\(.key) - kernel \(.value | .major_software.kernel[0]), released \(.value | .release_date)"' \
         | grep 'alpha tag'
```

Then you can use the "alpha tag" value with the `--sdk-version` option to install a custom SDK version.

#### Picking an OS image version from channels other than Alpha

> For adding new features and package updates we strongly recommend to base your work on an Alpha OS image version, which is what the above will set you up for. However, if you want to work on a small improvement or bug-fix based on a Beta or a Stable release, you will need to initialise your SDK correctly.

Reviewer: The above did not use Alpha as a base but only the Alpha SDK. I suggest to split this up into two parts. One is for using a particular release which also addresses the concern of changed ebuilds and mismatching binary package URLs. The other part can then cover using the maintenance branches `flatcar-MAJOR` (from `maintenance.xml`) or the `main` branch with a matching nightly binary package URL set up (This is what needs to be done if PRs will be filed).

Author: I'd rather have people base their work on Alpha (which has passed all tests) than on a nightly for the time being, and only include references on how to use nightlies in the "tips and tricks" doc.

Reviewer: Also good, but should we still cover the usage of development branches (maintenance branches and main) here with a reference to the nightly packages section of tips & tricks or move the usage of dev branches out, too?

Author: I'd rather put dev branches and nightlies into sdk-tips-and-tricks - I don't want to hide it, but I also want people to dig a bit to find it :) That said, do you think the respective [new section in tips-and-tricks](https://github.com/kinvolk/flatcar-docs/blob/28f7e723e053f14ea0d01e2c5482679255d5d4c3/docs/reference/developer-guides/sdk-tips-and-tricks.md#use-the-latest-nightly-build-sdk--flatcar-packages-from-the-main-branch) is exhaustive or should I elaborate there?

Then proceed with the installation of the cork binary to a location on your path:

```shell
$ chmod +x cork
$ mkdir -p ~/.local/bin
$ mv cork ~/.local/bin
$ export PATH=$PATH:$HOME/.local/bin
```

You may want to add the `PATH` export to your shell profile (e.g. `.bashrc`).

**Alternatively, clone the mantle repo and build cork from sources**

If you want to build cork from sources instead of using an official release, you'll need golang installed as mantle is written in go.

```shell
$ git clone git@github.com:flatcar-linux/mantle.git
$ cd mantle
$ ./build cork
```

> Now you want to make the resulting `bin/cork` binary part of your `$PATH`. Some prefer a symlink `~/.local/bin/cork -> ~/code/mantle/bin/cork ` ( `~/.local/bin/` often is in `$PATH` already as per `.bashrc`) - but you’re free to make your own arrangements as you see fit. Simply running `export PATH="$PATH:$(pwd/bin)"` in the mantle repo directory works (for the cuurent shell), too.

Reviewer: typo _cuurent_

$ git clone git@github.com:flatcar-linux/mantle.git
$ cd mantle
$ ./build cork
```

By default `cork create` will download the latest SDK used for the Alpha release which can be changed with `--sdk-version x.y.z`.
You can also change the SDK release later.
Now you want to make the resulting `bin/cork` binary part of your `$PATH`. Some prefer a symlink `~/.local/bin/cork -> ~/code/mantle/bin/cork ` ( `~/.local/bin/` often is in `$PATH` already as per `.bashrc`) - but you’re free to make your own arrangements as you see fit. Simply running `export PATH="$PATH:$(pwd/bin)"` in the mantle repo directory works (for the cuurent shell), too.

A manifest defines which source code branches are checked out. The default manifest uses the `main` development branches for the next Alpha major version.
If you want to check out a specific release, you can pass `--manifest-branch flatcar-CHANNEL-x.y.z`.
If you want to add changes to a major version that is propagated through the channels and gets maintenance releases increasing the minor version,
you need to use the maintenance branch `flatcar-MAJORVERSION` (the major version is the _x_ in _x.y.z_) which you can check out by passing

> `--manifest-branch flatcar-CHANNEL-x.y.z --manifest-name maintenance.xml` for any release of that major version family (not for releases <= 2513.2.0).

Reviewer: Please find a new place for this information, e.g., ` --manifest-name maintenance.xml` is not found anymore now and it's the quite useful because it checks out the right branches to file a patch for this release. Also the `main` vs `flatcar-MAJORVERSION` maintenance branches should be explained still.

(Using `--manifest-branch refs/tags/[TAG]` instead of `--manifest-branch flatcar-CHANNEL-x.y.z` is ok.)

Author: Thank you, good catch! `--manifest-name maintenance.xml` is now covered further down in the document (right before ["enter the SDK"](https://github.com/kinvolk/flatcar-docs/blob/28f7e723e053f14ea0d01e2c5482679255d5d4c3/docs/reference/developer-guides/sdk-modifying-flatcar.md#enter-the-sdk)) - the same section discusses the difference between using Alpha vs. a maintenance branch.
 
Documentation on how to use the main / nightly branches has been moved [to the sdk-tips-and-tricks](https://github.com/kinvolk/flatcar-docs/blob/28f7e723e053f14ea0d01e2c5482679255d5d4c3/docs/reference/developer-guides/sdk-tips-and-tricks.md#use-the-latest-nightly-build-sdk--flatcar-packages-from-the-main-branch) doc.

To reduce the build time it is important that as many binary packages as possible are downloaded from a previous SDK release,
Flatcar release, or nightly build.
For adding new features and package updates we strongly recommend to base your work on an Alpha OS image version, which is what the above will set you up for. However, if you want to work on a small improvement or bug-fix based on a Beta or a Stable release, you will need to initialise your SDK correctly.

To reference the latest nightly builds as source for board packages, use the following command. It has to be rerun if you later want
to switch to a recent nightly version because it only dereferences the nightly of the day when you ran this command, to keep your
build environment reproducible. For `BRANCH` use the branch you checked out or want to base your work on, e.g.,
the Alpha development branch `main` or the `flatcar-MAJORVERSION` maintenance branches.
To identify which tag to use, browse https://github.com/kinvolk/manifest/tags and find the version number of the release you want to create an SDK for. Then run 

```shell

> ./set_version --dev-board --board-version amd64-usr/BRANCH-nightly --dev-sdk --sdk-version sdk-main-nightly

Reviewer: Why where the two cases for running `./set_version` removed? It's helping people to fetch the binary packages and not waste hours due to unrelated compilations, please add it back.

In other words, you'll be dealing with two levels of chroot:
* The top-level SDK chroot, activated by `cork enter`, consisting of the build chroot with ebuild file sources (`src/` in the SDK directory) mounted in
* The image chroot, in `/]build/>arch>` in the SDK

Both chroots use Gentoo's portage to manage the packages: `sudo emerge` is used to manage SDK packages, and `emerge-<arch>` (`emerge-amd64-usr` or `emerge-arm64-usr`, without sudo) is used to do the same for the OS image roots.

#### Optional: Set a password for console login

Your SSH keys should be detected and added automatically by the image build process. Optionally, you can set a password for the `core` user which you can use later for ssh authentication, should SSH pubkey authentication not work for you.

After entering the chroot via `cork` for the first time, you can set user `core`'s password:

```shell
$ ./set_shared_user_password.sh
```

> This is the password you will use to log into the console of images built with the SDK.

Reviewer: Since we have `autologin` for most images, we can also omit this whole step here and more it to the tips & tricks.

    *   a platform independent part  - `commonconfig-<version>`
    *   platform dependent configs - `<arch>_defconfig-<version>`

The first section below will elaborate on developing, and testing, your modifications before we’ll merge into the ebuilds mentioned above. To achieve this, we will use the low level `ebuild` tool to selectively prepare the build of kernel image and modules, then make our modifications in Gentoo’s temporary build directories, and then continue the build. This way we’ll arrive at a boot-able, test-able image before merging your changes into the coreos-overlay ebuild file. Using Gentoo’s build-temp directories will also allow you to better iterate on your changes if you encounter problems during the build, or when testing your changes in a qemu image.

Only after we’ve tested our changes will we modify the kernel ebuild to persist the new configuration.

First, we will set up kernel and module sources, and modify those before build. To fetch and to configure the sources and to build a stock kernel, run:
```shell
$ cork enter

> $ ebuild-amd64-usr ../third_party/coreos-overlay/sys-kernel/coreos-modules/coreos-modules-<version>.ebuild configure

Reviewer: All these steps should maybe go to tips & ticks? Normally it's not needed to use them and often adding a kernel module is a one line change to the common config and the regular `./build_packages && ./build_image` flow to build the image.
